applications/OptimizationApplication/custom_utilities/container_variable_data_holder/container_variable_data_holder_base.h

```diff
@@ -0,0 +1,194 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   license: OptimizationApplication/license.txt
+//
+//  Main author:     Suneth Warnakulasuriya
+//
+
+#pragma once
+
+// System includes
+#include <string>
+
+// Project includes
+#include "includes/define.h"
+#include "includes/model_part.h"
+
+// Application includes
+
+namespace Kratos {
+
+///@name Kratos Classes
+///@{
+
+class HistoricalContainerDataIO
+{
+public:
+    ///@name Public operations
+    ///@{
+
+    template<class TDataType>
+    static TDataType& GetValue(
+        ModelPart::NodeType& rNode,
+        const Variable<TDataType>& rVariable)
+    {
+        return rNode.FastGetSolutionStepValue(rVariable);
+    }
+
+    template<class TDataType>
+    static void SetValue(
+        ModelPart::NodeType& rNode,
+        const Variable<TDataType>& rVariable,
+        const TDataType& rValue)
+    {
+        rNode.FastGetSolutionStepValue(rVariable) = rValue;
+    }
+
+    ///@}
+};
```

Comment:
Getting and setting data from/to different sources/destinations like this is always painful, precisely because the names are different. Here, you provide a uniform naming for the functions (`SetValue`, `GetValue`) but still give different names to the classes, so the same problem is propagated.

I'd suggest putting a stop to naming historical and nonhistorical data operations, and make the distinction in the template arguments instead, by introducing tags for the different data types. For example:

```cpp
namespace Tags {
    struct Historical {};
    struct NonHistorical {};
} // namespace Tags

// Dummy generic class to be specialized later
template <class TTag>
struct ContainerDataIO {};

template <>
struct ContainerDataIO<Tags::Historical>
{
    template <class TData>
    static const TData& GetValue(const ModelPart::NodeType& rNode,
                                 const Variable<TData>& rVariable)
    {
        return rNode.FastGetSolutionStepValue(rVariable);
    }

    ...
}; // struct ContainerDataIO<Tags::Historical>

template <>
struct ContainerDataIO<Tags::NonHistorical>
{
    template<class TData, class TEntity>
    static const TData& GetValue(
        const TEntity& rEntity,
        const Variable<TData>& rVariable)
    {
        return rEntity.GetValue(rVariable);
    }

    ...
}; // struct ContainerDataIO<Tags::NonHistorical>
```

It'll be more template-friendly later on if you don't have to worry about different names.

Reply:
done :)